wcfsetup/install/files/acp/update_com.woltlab.wcf_5.4_migrate_multifactor.php

```diff
@@ -0,0 +1,151 @@
+use wcf\system\user\multifactor\Setup;
+use wcf\system\WCF;
+
+$hanashiTwoStep = PackageCache::getInstance()->getPackageByIdentifier('eu.hanashi.wsc.two-step-verification');
+
+if (!$hanashiTwoStep) {
+	return;
+}
+
+// Fetch the object types for the relevant MFA methods.
+$totpMethod = ObjectTypeCache::getInstance()->getObjectTypeByName('com.woltlab.wcf.multifactor', 'com.woltlab.wcf.multifactor.totp');
+$backupMethod = ObjectTypeCache::getInstance()->getObjectTypeByName('com.woltlab.wcf.multifactor', 'com.woltlab.wcf.multifactor.backup');
+
+// Fetch the backup code hashing algorithm.
+// We use the Wcf1 algorithm as it's super cheap compared to BCrypt and the previous
+// backup codes were stored in plaintext, leading to a net improvement.
+$hashAlgorithm = new Wcf1();
+$hashAlgorithmName = PasswordAlgorithmManager::getInstance()->getNameFromAlgorithm($hashAlgorithm);
+
+// Fetch the affected user IDs.
+$sql = "SELECT	DISTINCT userID
+	FROM	wcf".WCF_N."_user_authenticator
+	WHERE		type = ?
+		AND	userID NOT IN (
+			SELECT	userID
+			FROM 	wcf".WCF_N."_user_multifactor
+			WHERE	objectTypeID = ?
+		)";
+$statement = WCF::getDB()->prepareStatement($sql);
+$statement->execute([
+	'totp',
+	$totpMethod->objectTypeID,
+]);
+$userIDs = $statement->fetchAll(\PDO::FETCH_COLUMN);
+
+// Prepare the statements for use in user processing.
+$sql = "SELECT	name, secret, time
+	FROM	wcf".WCF_N."_user_authenticator
+	WHERE		type = ?
+		AND	userID = ?
+	FOR UPDATE";
+$existingTotpAuthenticatorStatement = WCF::getDB()->prepareStatement($sql);
+$sql = "SELECT	backupCode
+	FROM	wcf".WCF_N."_user_backup_code
+	WHERE	userID = ?
+	FOR UPDATE";
+$existingBackupStatement = WCF::getDB()->prepareStatement($sql);
+
+$sql = "INSERT INTO	wcf".WCF_N."_user_multifactor_totp
+			(setupID, deviceID, deviceName, secret, minCounter, createTime)
+	VALUES		(?, ?, ?, ?, ?, ?)";
+$createTotpStatement = WCF::getDB()->prepareStatement($sql);
+$sql = "INSERT INTO	wcf".WCF_N."_user_multifactor_backup
+			(setupID, identifier, code, createTime)
+	VALUES		(?, ?, ?, ?)";
+$createBackupStatement = WCF::getDB()->prepareStatement($sql);
+
+// TODO: Do we need to split this across multiple requests?
+foreach ($userIDs as $userID) {
+	WCF::getDB()->beginTransaction();
+	
+	// Do not use UserRuntimeCache due to possible memory constraints.
+	$user = new User($userID);
+	$userEditor = new UserEditor($user);
+	
+	if (Setup::find($totpMethod, $user) !== null) {
+		// Skip this user, because they have an enabled TOTP method.
+		// This should never happen, because these users are filtered out
+		// when selecting, but we are going to play safe.
+		continue;
+	}
+	
+	$totpSetup = Setup::allocateSetUpId($totpMethod, $user);
+	
+	$existingTotpAuthenticatorStatement->execute([
+		'totp',
+		$user->userID,
+	]);
+	$earliestTotp = null;
+	while ($row = $existingTotpAuthenticatorStatement->fetchArray()) {
+		$createTotpStatement->execute([
+			$totpSetup->getId(),
+			\bin2hex(\random_bytes(16)),
```

Comment:
Maybe use `Hex::encode()` for consistency reasons.

Reply:
I'm intentionally not doing this here for 2 reasons:

1) It is not necessary, because this is a privileged script that cannot be triggered by an unauthenticated 3rd party.
2) I want to keep the update scripts free of complex libraries for resiliency.

Reply:
I realize that we need the library anyway, so I'll adjust this.